openedx/core/djangoapps/theming/views.py

```diff
@@ -0,0 +1,104 @@
+"""
+Views file for the Darklang Django App
+"""
+
+from django.template.loader import render_to_string
+from django.utils.translation import ugettext as _
+from openedx.core.djangoapps.plugin_api.views import EdxFragmentView
+from openedx.core.djangoapps.user_api.preferences.api import (
+    delete_user_preference,
+    get_user_preference,
+    set_user_preference,
+)
+from openedx.core.djangoapps.util.user_messages import (
+    register_error_message,
+    register_success_message,
+)
+from web_fragments.fragment import Fragment
+
+from .helpers import theme_exists
+from .models import SiteTheme
+
+PREVIEW_SITE_THEME_PREFERENCE_KEY = 'preview-site-theme'
+
+
+def get_user_preview_site_theme(request):
+    """
+    Returns the preview site for the current user, or None if not set.
+    """
+    user = request.user
+    if not user or user.is_anonymous():
+        return None
+    preview_site_name = get_user_preference(user, PREVIEW_SITE_THEME_PREFERENCE_KEY)
+    if not preview_site_name:
+        return None
+    return SiteTheme(site=request.site, theme_dir_name=preview_site_name)
+
+
+def set_user_preview_site_theme(request, preview_site_theme):
+    """
+    Sets the current user's preferred preview site theme.
+
+    Args:
+        request: the current request
+        preview_site_theme: the preview site theme (or None to remove it)
+    """
+    if preview_site_theme:
+        if isinstance(preview_site_theme, SiteTheme):
+            preview_site_theme_name = preview_site_theme.theme_dir_name
+        else:
+            preview_site_theme_name = preview_site_theme
+        if theme_exists(preview_site_theme_name):
+            set_user_preference(request.user, PREVIEW_SITE_THEME_PREFERENCE_KEY, preview_site_theme_name)
+            register_success_message(
+                request,
+                _('Site theme changed to {site_theme}'.format(site_theme=preview_site_theme_name))
+            )
+        else:
+            register_error_message(
+                request,
+                _('Theme {site_theme} does not exist'.format(site_theme=preview_site_theme_name))
+            )
+    else:
+        delete_user_preference(request.user, PREVIEW_SITE_THEME_PREFERENCE_KEY)
+        register_success_message(request, _('Site theme reverted to the default'))
+
+
+class ThemingAdministrationFragmentView(EdxFragmentView):
+    """
+    Fragment view to allow a user to administer theming.
+    """
+
+    def render_to_fragment(self, request, course_id=None, **kwargs):
```

Comment:
In this case, the challenge was how to get a page that renders with two different templates, i.e. main.html in LMS and base.html in Studio. The solution that the update_lang page used seemed like a hack to me, while it occurred to me that a fragment was the perfect solution. A Django app in openedx can provide a fragment, and then it will render with a different wrapper in the LMS and Studio. Every fragment view provides a 'render as standalone' option, so I could benefit from that here. I'm not sure how best to communicate this. I'll start by adding this as an example to this wiki page: https://openedx.atlassian.net/wiki/display/FEDX/Web+Fragments